pytorch_lightning/callbacks/progress/base.py

```diff
@@ -47,52 +47,57 @@ def on_train_batch_end(self, trainer, pl_module, outputs, batch_idx):
     """
 
     def __init__(self):
-
-        self._trainer = None
-        self._train_batch_idx = 0
-        self._val_batch_idx = 0
-        self._test_batch_idx = 0
-        self._predict_batch_idx = 0
+        self._trainer: Optional["pl.Trainer"] = None
 
     @property
-    def trainer(self):
+    def trainer(self) -> "pl.Trainer":
         return self._trainer
 
     @property
     def train_batch_idx(self) -> int:
-        """The current batch index being processed during training.
+        """The number of batches processed during training.
 
         Use this to update your progress bar.
         """
-        return self._train_batch_idx
+        if self.trainer is None:
+            return 0
+        return self.trainer.fit_loop.epoch_loop.batch_progress.current.processed
 
     @property
     def val_batch_idx(self) -> int:
-        """The current batch index being processed during validation.
+        """The number of batches processed during validation.
 
         Use this to update your progress bar.
         """
-        return self._val_batch_idx
+        if self.trainer is None:
+            return 0
+        if self.trainer.state.fn == "fit":
+            return self.trainer.fit_loop.epoch_loop.val_loop.epoch_loop.batch_progress.current.processed
+        return self.trainer.validate_loop.epoch_loop.batch_progress.current.processed
 
     @property
     def test_batch_idx(self) -> int:
-        """The current batch index being processed during testing.
+        """The number of batches processed during testing.
 
         Use this to update your progress bar.
         """
-        return self._test_batch_idx
+        if self.trainer is None:
+            return 0
+        return self.trainer.test_loop.epoch_loop.batch_progress.current.processed
```

Comment:
Did you choose this over `self.trainer.fit_loop.batch_idx` as you are only want to display the processed value?

Reply:
This is the `test_batch_idx` so we wouldn't access the `fit_loop.batch_idx`

And `processed` is what matches the previous manually-managed value when the hooks are called.